flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java

```diff
@@ -473,25 +519,41 @@ public T deserialize(T reuse, DataInputView source) throws IOException {
             }
         }
 
-        if ((flags & NO_SUBCLASS) != 0) {
+        if (isRecord) {
```

Comment:
Correct me here, if I'm wrong but I have the impression that `deserialize(DataInputView source)` can be implemented in the following way:
```java
public T deserialize(DataInputView source) throws IOException {
    deserialize(null, source);
}
```
It's a bit tricky to investigate because if conditions are differently implemented. But the code paths match with `reuse == null` as far as I can see. This would reduce the amount of code redundancy in this class tremendously (it might deserve its own commit, I guess? :thinking: ).

The same seems to be true for the `copy` command. Refactoring that would reduce the diff by quite a bit as well. WDYT?

Reply:
tbh, I feel like the `TypeSerializer` interface could be cleaned up in this regard. Some `reuse` should be marked as `@Nullable` and each implementation should be able to handle that specific case (i.e. instantiating a new instance) rather than having two separate implementations. There would be a default implementation of `copy(T)` and `deserialize(T)` in that case. But that's out-of-scope for this PR.

Reply:
I will improve the deserialise but the copy is a bit trickier.

Reply:
Your suspicion/caution was correct - it's actually not that straight-forward setting `reuse = null` due to the recursive call of `deserialize(DataInputView)`/`deserialize(T, DataInputView)` on the fields. -.-

Anyway, I went through the code once more to figure out how we could reduce the redundant code here a bit more:
```java
    private void deserializeFields(
            DataInputView source,
            BiConsumerWithException<Integer, Object, IllegalAccessException> fieldInitializer)
            throws IOException {
        deserializeFields(
                source,
                fieldInitializer,
                fieldIdx -> fieldSerializers[fieldIdx].deserialize(source));
    }

    private void deserializeFields(
            DataInputView source,
            T reuse,
            BiConsumerWithException<Integer, Object, IllegalAccessException> fieldInitializer)
            throws IOException {
        deserializeFields(
                source,
                fieldInitializer,
                fieldIdx -> {
                    final Object reuseFieldValue =
                            reuse == null ? null : fields[fieldIdx].get(reuse);
                    if (reuseFieldValue != null) {
                        return fieldSerializers[fieldIdx].deserialize(reuseFieldValue, source);
                    } else {
                        return fieldSerializers[fieldIdx].deserialize(source);
                    }
                });
    }

    private void deserializeFields(
            DataInputView source,
            BiConsumerWithException<Integer, Object, IllegalAccessException> fieldInitializer,
            FunctionWithException<Integer, Object, ? extends Exception> valueSupplier)
            throws IOException {
        try {
            for (int i = 0; i < numFields; i++) {
                boolean isNull = source.readBoolean();

                if (fields[i] != null && isNull) {
                    fieldInitializer.accept(i, null);
                } else if (fields[i] != null && !isNull) {
                    fieldInitializer.accept(i, valueSupplier.apply(i));
                } else if (fields[i] == null && !isNull) {
                    // read and dump a pre-existing field value
                    fieldSerializers[i].deserialize(source);
                }
            }
        } catch (IllegalAccessException e) {
            throw new RuntimeException(
                    "Error during POJO copy, this should not happen since we check the fields before.",
                    e);
        } catch (Throwable e) {
            ExceptionUtils.rethrowIOException(e);
        }
    }

```

...just as a proposal. I just saw your comment that your not too excited about using consumers. I see your point there and I'm not pushing for it. I wanted to share my findings anyway.

Reply:
oh yea you are right, I will actually revery my deserialise refactor, to avoid this unintended change. 

I will not do this refactoring now. The consumers in theory can solve this but I suspect there will be an intended performance hit. I also don't want to spend investigating that too much. Will open a follow up ticket.